### PR TITLE
Allow implicit casting and details for literals.

### PR DESCRIPTION
**Pull request by mbhealy**
### Summary
This PR is the result of discussions among the members of the Types and Casting Work Group, and introduces more detail in the spec regarding the rules of implicit casting. In general, classical standard types mimic the C99 spec. Angles and cbits do not implicitly cast to the standard types and vice versa, however, angles will promote/degrade to target angles with differing precision.

I've also included more detail on the format for literals in the language.

### Details and comments
This PR supercedes #326

### CODE REVIEW

**Review by jwoehr — CHANGES_REQUESTED** [review]
### `source/language/types.rst`
>There are three types of literals in OpenQASM 3, integer, float, and bit string.

1. The [grammar](https://github.com/Qiskit/openqasm/blob/main/source/grammar/qasm3.g4) also defines `booleanLiteral` and `TimingLiteral`.
2. The grammar defines `bit` and `bitType` but neither the grammar nor the text mention _bit string_. I'm not sure the author's intent here is to introduce terminology without definition. 

>Float literals contain a digit followed by a ``.``, a ``.`` followed by a digit, or use scientific notation, and may be given an explicit sign.

This seems slightly at odds with the grammar:
```
fragment SciNotation : [eE] ;
fragment PlusMinus : PLUS | MINUS ;
fragment Float : Digit+ DOT Digit* ;
RealNumber : (Integer | Float ) (SciNotation PlusMinus? Integer )? ;
```
which describes a `Float` as one _or more_ digits, a 'dot', followed by _zero or more_ digits. _Also, the existence of a prepended sign is not obvious to me in the grammar._

>Bit string literals are denoted by double quotes ``"`` surrounding a number of zero and one digits, and may include non-consecutive underscores to improve readability for large strings.

I think this requires a change to the grammar.

>The standard types follow rules that mimick those of C99 for `promotion and conversion <https://en.cppreference.com/w/c/language/conversion>`

Do we really want to wave at the C standard in lieu of spelling this out explicitly?

**Comment by mbhealy** [comment]
The grammar still needs to be updated to match the new literal definitions provided in the text.

**Comment by jwoehr** [comment]
> The grammar still needs to be updated to match the new literal definitions provided in the text.
Is that part of this PR or a separate one?

**Comment by mbhealy** [comment]
> Is that part of this PR or a separate one?

The grammar will be updated in this PR.

**Comment by mbhealy** [comment]
> Do we really want to wave at the C standard in lieu of spelling this out explicitly?

Personally this is my preference, but I am open to trying to spell it out if that is the consensus view. In the discussion during the TSC meeting @levbishop indicated he agreed with me that pointing to a fully fleshed-out spec was a reasonable way to build on other's work.

**Comment by jwoehr** [comment]
> > Do we really want to wave at the C standard in lieu of spelling this out explicitly?
> 
> Personally this is my preference, but I am open to trying to spell it out if that is the consensus view. In the discussion during the TSC meeting @levbishop indicated he agreed with me that pointing to a fully fleshed-out spec was a reasonable way to build on other's work.

Yes, that was the conversation. Believe me, I'm not trying to raise an ideological point, nor am I seeking the fast path to endless debate. I'm just reading the [cited work](https://en.cppreference.com/w/c/language/conversion) and seeing how rich it is and potentially difficult to interpret in the context of OpenQASM. What's an `angle` to the C standard, or a function-to-pointer conversion to OpenQASM?

In our necessarily rapid march to 3.0 it may be necessary patchwork, but by 3.1 we will have to be spell it out, if for no other reason than that deferring to another language's standard may offer leverage to any hostile critics.

**Review by jwoehr — COMMENTED** [review]
I believe I agree, @mbhealy , but just confirming to be sure that there is no longer (or never really was) any entity in OpenQASM 3 such as a `StringLiteral` meaning an invariable string of language characters, and that the only thing bearing any such-sounding name is really the `BitStringLiteral` for representing binary bit values ... Correct?

**Comment by jwoehr** [comment]
Other than the above two comments/concerns, all my points are addressed by the recent commits.
- the grammar looks good
- the test cases look good
- the text looks good

**Comment by mbhealy** [comment]
> I believe I agree, @mbhealy , but just confirming to be sure that there is no longer (or never really was) any entity in OpenQASM 3 such as a `StringLiteral` meaning an invariable string of language characters, and that the only thing bearing any such-sounding name is really the `BitStringLiteral` for representing binary bit values ... Correct?

Correct, the grammar still contains a StringLiteral, but it is only used for `include` statements, and for the name of the pulse grammar. 

The only strings allowed generically within a program are BitStringLiterals.

**Comment by mbhealy** [comment]
> What's an `angle` to the C standard, or a function-to-pointer conversion to OpenQASM?

I specifically waive to the C standard *only* for the standard types, float, int/uint, and bool. We also have no pointers of any type, so there is no confusion there. Angles, durations, and bit-strings must be explicitly cast to a standard type and vice versa.

**Comment by jwoehr** [comment]
@mbhealy

1. w/r/t `StringLiteral` .. Tt's still there for those two cases `include` and pulse grammar name? I was just rapid-eye-scanning your changes and somehow got the impression from the test cases `StringLiteral` had disappeared in favor of `BitStringLiteral`, but they are 2 separate things now (he asks)?
2. w/r/t C-style implicit conversions vs. the list of explicitly cast items ... as long we're clearly saying it.

So granted the above, I have no further comment.

**Comment by mbhealy** [comment]
> 1. w/r/t `StringLiteral` .. Tt's still there for those two cases `include` and pulse grammar name? I was just rapid-eye-scanning your changes and somehow got the impression from the test cases `StringLiteral` had disappeared in favor of `BitStringLiteral`, but they are 2 separate things now (he asks)?

Yes, `StringLiteral` is no longer an expressionTerminator. It may only occur between the `include` token and a `;` or between the `defcalgrammar` token and a `;`.

`BitStringLiteral` *is* an `expressionTerminator`, so it can occur in many more places where it does not make sense to accept a generic string.

File: source/openqasm/openqasm3/parser.py
Comment (psivaraj) [inline source/openqasm/openqasm3/parser.py:579]: **question**: Should we strip the `_` here?

File: source/language/types.rst
Comment (psivaraj) [inline source/language/types.rst:644]: **suggestion**: Point folks to the portion of the spec where they can find these.

File: source/language/types.rst
Comment (psivaraj) [inline source/language/types.rst:646]: **suggestion**: Point folks to the portion of the spec where they can find more information.

File: source/grammar/qasm3.g4
Comment (mbhealy) [reply source/grammar/qasm3.g4]: Good suggestion, added in https://github.com/Qiskit/openqasm/pull/337/commits/b6dfa2e7498f98c80d7fe0d2875480354c497a3b

File: source/language/types.rst
Comment (mbhealy) [reply source/language/types.rst]: Excellent suggestion, added in https://github.com/Qiskit/openqasm/pull/337/commits/c1304adc2bc2341fbf0d46ae200a8c6b76ce7a11

File: source/language/types.rst
Comment (mbhealy) [reply source/language/types.rst:335]: Done in https://github.com/Qiskit/openqasm/pull/337/commits/7d4a67a42af923052a04122269e99d3638a8adaf

File: source/openqasm/openqasm3/parser.py
Comment (mbhealy) [reply source/openqasm/openqasm3/parser.py]: Good point, added in https://github.com/Qiskit/openqasm/pull/337/commits/dc51d0df2e4ddc61a9fa8de21df9f162bb9282c9

File: source/language/types.rst
Comment (mbhealy) [reply source/language/types.rst]: Added in https://github.com/Qiskit/openqasm/pull/337/commits/d7117818df3033cb7b8e7ec9f556d132ac7b0fbe

File: source/language/types.rst
Comment (mbhealy) [reply source/language/types.rst:653]: Added in https://github.com/Qiskit/openqasm/pull/337/commits/8f455d32344aa1e3513944180224ab5e4bff1b91

File: source/language/types.rst
Comment (jakelishman) [inline source/language/types.rst:644]:
```suggestion
The standard types follow rules that mimic those of C99 for `promotion and
```

File: source/language/types.rst
Comment (jakelishman) [reply source/language/types.rst:655]: +1 - I think there's a few cases where we differ from C99 in this respect, so it'd definitely be good to put in a cross-reference.

**Review by jakelishman — COMMENTED** [review]
This looks like a good addition, thanks, especially with regard to the additional information about literals.

File: source/language/types.rst
Comment (mbhealy) [reply source/language/types.rst:655]: We discussed this in the WG today and realized that the main places we would point to are part of Jake's branch in his private repo specifying all of the operators and their input and output types, so if this PR gets merged before his he will do the necessary linking.

File: source/language/types.rst
Comment (mbhealy) [reply source/language/types.rst]: Good point, changed in https://github.com/Qiskit/openqasm/pull/337/commits/18c03f3c591aabf8aa6e3cd56cb4ddeb211e092f

File: source/language/types.rst
Comment (mbhealy) [reply source/language/types.rst]: Changed in https://github.com/Qiskit/openqasm/pull/337/commits/e6e1c2835d4b57e656f9fc141f3a9187db210930

File: source/grammar/qasm3.g4
Comment (jakelishman) [inline source/grammar/qasm3.g4:649]: The addition of `DecIntegerLiteral` makes some things a bit weird because it's not exactly the same as `Integer`.  All `Integer`s are valid `DecIntegerLiteral`s (but not vice-versa), and because `Integer` is defined first, ANTLR will always tokenise to `Integer` preferentially.  To avoid weird duplication in `expressionTerminator`, I think we might want to:
- make `Integer` a `fragment` rule
- remove `Integer` from `expressionTerminator`
- rewrite the `version` rule
- remove the `Integer` handling from the reference parser's `visitExpressionTerminator` (since the specific literals will now handle it)

The last point is currently tricky to do right (e.g. it currently allows `RealNumber` - `OPENQASM 3e+0;`!), so for now, I'd just set
```antlr
version: 'OPENQASM' DecIntegerLiteral ('.' DecIntegerLiteral)? SEMICOLON;
```
(which isn't ideal because it'd technically allow underscores in the version number).  Once #278 is in place, I can fix things in the reference implementation in a very neat way to tokenise the version number better (including potential alpha/beta/rc suffixes) without messing with the rest of the tokenisation.

**Review by jakelishman — COMMENTED** [review]
All the spec changes look good to me, since we decided to go for more weak typing (though I'm internally sad we're leaving strong0 typing).  Thanks for taking care of all the reference grammar and AST+parser stuff as well - it all looks very neat and easy to read - I just have one comment about some odd duplication with `Integer`/`DecIntegerLiteral` that I can help with if needs be.

**Comment by jakelishman** [comment]
This PR is now rebased over other changes to the grammar, and I believe all the current textual comments on the spec are now either resolved, or have been punted to other PRs.

@jwoehr: are you happy your comments have been addressed?  It looks like it above, but I just wanted to check.

**Review by jwoehr — APPROVED** [review]

**Comment by jwoehr** [comment]
If we're all happy, I will push the Merge button!
All good?!

**Comment by jakelishman** [comment]
I don't think this one has TSC approval yet but it needs it, so we need to wait for that.

We need TSC review [on the RST files in the PR](337/files?file-filters%5B%5D=.rst&show-viewed-files=true), but the other changes are just code-related.

**Review by stevenheidel — APPROVED** [review]

**Review by blakejohnson — DISMISSED** [review]

**Review by blakejohnson — APPROVED** [review]
Re-approving.

**Pull request merged**
